server/src/test/java/org/apache/kafka/api/EndToEndClusterIdTest.java

```diff
@@ -0,0 +1,215 @@
+import org.apache.kafka.common.test.api.ClusterTestDefaults;
+import org.apache.kafka.server.metrics.MetricConfigs;
+import org.apache.kafka.test.MockConsumerInterceptor;
+import org.apache.kafka.test.MockDeserializer;
+import org.apache.kafka.test.MockMetricsReporter;
+import org.apache.kafka.test.MockProducerInterceptor;
+import org.apache.kafka.test.MockSerializer;
+import org.apache.kafka.test.TestUtils;
+
+import org.junit.jupiter.api.BeforeEach;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static org.apache.kafka.test.TestUtils.isValidClusterId;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+
+@ClusterTestDefaults(serverProperties = {
+    @ClusterConfigProperty(key = MetricConfigs.METRIC_REPORTER_CLASSES_CONFIG, value = "org.apache.kafka.api.EndToEndClusterIdTest$MockBrokerMetricsReporter"),
+})
+public class EndToEndClusterIdTest {
+
+    private static final String TOPIC = "e2etopic";
+    private static final int PARTITION = 0;
+    private static final int NUM_RECORDS = 1;
+    private static final TopicPartition TP = new TopicPartition(TOPIC, PARTITION);
+    private final ClusterInstance clusterInstance;
+
+    EndToEndClusterIdTest(ClusterInstance clusterInstance) {
+        this.clusterInstance = clusterInstance;
+    }
+
+    @BeforeEach
+    public void setup() throws InterruptedException {
+        this.clusterInstance.createTopic(TOPIC, 2, (short) 1);
+        MockDeserializer.resetStaticVariables();
+    }
+
+    public static class MockConsumerMetricsReporter extends MockMetricsReporter implements ClusterResourceListener {
+        public static final AtomicReference<ClusterResource> CLUSTER_META = new AtomicReference<>();
+
+        @Override
+        public void onUpdate(ClusterResource clusterMetadata) {
+            CLUSTER_META.set(clusterMetadata);
+        }
+    }
+
+    public static class MockProducerMetricsReporter extends MockMetricsReporter implements ClusterResourceListener {
+        public static final AtomicReference<ClusterResource> CLUSTER_META = new AtomicReference<>();
+
+        @Override
+        public void onUpdate(ClusterResource clusterMetadata) {
+            CLUSTER_META.set(clusterMetadata);
+        }
+    }
+
+    public static class MockBrokerMetricsReporter extends MockMetricsReporter implements ClusterResourceListener {
+        public static final AtomicReference<ClusterResource> CLUSTER_META = new AtomicReference<>();
+
+        @Override
+        public void onUpdate(ClusterResource clusterMetadata) {
+            CLUSTER_META.set(clusterMetadata);
+        }
+    }
+
+    @ClusterTest
+    public void testEndToEndWithClassicProtocol(ClusterInstance clusterInstance) throws Exception {
```

Comment:
Please remove `ClusterInstance clusterInstance` parameter, we don't use it in the test.